src/explore-education-statistics-common/src/modules/table-tool/utils/mapTableToJson.ts

```diff
@@ -0,0 +1,157 @@
+*/
+export default function mapTableToJson(
+  tableHeadersConfig: TableHeadersConfig,
+  subjectMeta: FullTableMeta,
+  results: TableDataResult[],
+  query?: ReleaseTableDataQuery,
+): {
+  tableJson: TableJson;
+  showMissingRowsOrColumnsWarning?: boolean;
+} {
+  const rowHeadersCartesian = cartesian(
+    ...tableHeadersConfig.rowGroups,
+    tableHeadersConfig.rows,
+  );
+
+  const columnHeadersCartesian = cartesian(
+    ...tableHeadersConfig.columnGroups,
+    tableHeadersConfig.columns,
+  );
+
+  // Track which columns actually have text values
+  // as we want to remove empty ones later.
+  const columnsWithText = columnHeadersCartesian.map(() => false);
+
+  const tableHeaderFilters = [
+    ...tableHeadersConfig.columnGroups.flatMap(filterGroup => filterGroup),
+    ...tableHeadersConfig.rowGroups.flatMap(filterGroup => filterGroup),
+    ...tableHeadersConfig.columns,
+    ...tableHeadersConfig.rows,
+  ].map(filter => filter.id);
+
+  const excludedFilters = getExcludedFilters(tableHeaderFilters, subjectMeta);
+
+  // Group measures by their respective combination of filters
+  // allowing lookups later on to be MUCH faster.
+  const measuresByFilterCombination = groupResultMeasuresByCombination(
+    results,
+    excludedFilters,
+  );
+
+  // Create a cartesian table of cells
+  const tableCartesian = makeTableCartesian(
+    rowHeadersCartesian,
+    columnHeadersCartesian,
+    measuresByFilterCombination,
+    columnsWithText,
+  );
+
+  const filteredCartesian = tableCartesian
+    .filter(row => row.some(cell => cell.text !== EMPTY_CELL_TEXT))
+    .map(row => row.filter((_, index) => columnsWithText[index]));
+
+  const rowHeaders = filteredCartesian.reduce<Header[]>((acc, row) => {
+    // Only need to use first column's rowFilters
+    // as they are the same for every column.
+
+    const filters = optimizeFilters(row[0].rowFilters, [
+      ...tableHeadersConfig.rowGroups,
+      tableHeadersConfig.rows,
+    ]);
+
+    return addFilters(acc, filters);
+  }, []);
+
+  // Only need to use first row's columnFilters
+  // as they are the same for every row.
+  const columnHeaders = filteredCartesian[0].reduce<Header[]>((acc, column) => {
+    const filters = optimizeFilters(column.columnFilters, [
+      ...tableHeadersConfig.columnGroups,
+      tableHeadersConfig.columns,
+    ]);
+
+    return addFilters(acc, filters);
+  }, []);
+
+  const rows = filteredCartesian.map(row => row.map(cell => cell.text));
+
+  // We 'expand' our headers so that we create the real table
+  // cells we need to render in array format (instead of a tree).
+
+  const expandedColumnHeaders = createExpandedColumnHeaders(columnHeaders);
+
+  const expandedRowHeaders = createExpandedRowHeaders(rowHeaders);
+
+  const totalColumns = sumBy(expandedRowHeaders[0], header => header.crossSpan);
+
+  const mappedColHeaders = mapColumns(expandedColumnHeaders, totalColumns);
+
+  const mappedRows = mapRows(rows, expandedRowHeaders);
```

Comment:
Would be tempted to rename the variables and functions to the following to make things clearer:

```ts
const thead = mapTableHead(expandedColumnHeaders, totalColumns);
const tbody = mapTableBody(rows, expandedRowHeaders);

return {
  tableJson: {
    thead,
    tbody,
  }
}
```